### PR TITLE
`Storable` trait and `Mapping`

### PR DESCRIPTION
**Pull request by mohammadfawaz**
🚨  THIS IS A DRAFT FOR EARLY FEEDBACK ONLY AND IS FAR FROM DONE 🚨 

Meant to close https://github.com/FuelLabs/sway/issues/1399 and also implement `Mapping`.

Blockers:
* [ ] https://github.com/FuelLabs/sway/issues/1159 to be able to see `impl<K, V> Mapping<K, V> where V: Storable`
* [ ] https://github.com/FuelLabs/sway/issues/1143 to be able to have `Mapping` to not need to use `K` and `V` in its data members.

### CODE REVIEW

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs [hunk 1]
```diff
@@ -516,6 +519,16 @@ impl TypedExpression {
                 },
                 span,
             ),
+            Expression::BuiltinGenerateB256Seed { span } => ok(
```
Comment (emilyaherbert) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs:522]: What about writing the equivalent of "GenerateB256Seed" as a function in the standard library? That way we don't have to special case it in the compiler.

**Review by emilyaherbert — CHANGES_REQUESTED** [review]

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs (hunk 1 above)
Comment (mohammadfawaz) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs:522]: The compiler passes through the intrinsic `__generate_b256_seed` (name is WIP) all the way to codegen where the actual generation happen. I'm not sure we can do that with pure library code and without a new intrinsic, right?

**Comment by otrho** [comment]
Initial thoughts: I don't mind generating the unique keys in an intrinsic, but I don't think it actually needs to go all the way to IR?  It looks like everything that is new in `from_ir.rs` could just go in `optimize.rs`, i.e., the keys are generated during IR generation, and just stored in the data section.  Is this not true?

And so I'm understanding correctly, the `Mapping` is essentially a hash map (with `insert(key, value)` and `get(key)`)
and a wrapper around the storage key management, so that the user doesn't have to think about the key?  And it would be generic if we had support in the compiler, but for now it's just `u64` -> `u64`?  I'm not really a fan of the name `Mapping`, even just `Map` or `HashMap` would be better IMO.

Otherwise, looking good! :+1:

**Comment by mohammadfawaz** [comment]
> Is this not true?
So this is the part that gets ugly... `new()` needs to be inlined for this to work because before inlining, we only have a single copy of this intrinsic (by looking at the IR before inlining for example). Or are you saying we can do traversal of the AST to figure out when to increment the counter by chasing `FunctionApplication`s?

>  so that the user doesn't have to think about the key
Yup!

> And it would be generic if we had support in the compiler, but for now it's just `u64` -> `u64`?
Yeah! there are a few blockers with generics right now. Just thought I'd showcase `u64 -> u64` first to get feedback on the overall approach, though I don't expect any of this to go in before we make it generic.

> I'm not really a fan of the name `Mapping`, even just `Map` or `HashMap` would be better IMO.
Potentially yeah. `Mapping` is from Solidity really and maybe `Map` is a better option. We can bikeshed over the name in the coming days.

**Comment by otrho** [comment]
> So this is the part that gets ugly... `new()` needs to be inlined for this to work because before inlining, we only have a single copy of this intrinsic

I see.  Yuck.  I would be _slightly_ more comfortable calling it `generate_uid()` since that sounds more general and low level, like something that belongs in IR. :wink:

**Pull request closed without merging**